### ISSUE TITLE
Allow serializing enums to integers in JSON output

### ISSUE THREAD

**Issue by ToluwaniO** [stephenh/ts-proto#96]
Is it possible to add a build option to serialize enums to integers instead of strings?
Here's what the library currently generates
```
export const ApprovalState = {
  APPROVED: 0 as const,
  PENDING: 1 as const,
  DENIED: 2 as const,
  UNRECOGNIZED: -1 as const,
  fromJSON(object: any): ApprovalState {
    switch (object) {
      case 0:
      case "APPROVED":
        return ApprovalState.APPROVED;
      case 1:
      case "PENDING":
        return ApprovalState.PENDING;
      case 2:
      case "DENIED":
        return ApprovalState.DENIED;
      case -1:
      case "UNRECOGNIZED":
      default:
        return ApprovalState.UNRECOGNIZED;
    }
  },
  toJSON(object: ApprovalState): string {
    
switch (object) {
      case ApprovalState.APPROVED:
        return "APPROVED";
      case ApprovalState.PENDING:
        return "PENDING";
      case ApprovalState.DENIED:
        return "DENIED";
      default:
        return "UNKNOWN";
    }
  },
}
```

Here's what I'm trying to achieve
```
export const ApprovalState = {
  APPROVED: 0 as const,
  PENDING: 1 as const,
  DENIED: 2 as const,
  UNRECOGNIZED: -1 as const,
  fromJSON(object: any): ApprovalState {
    switch (object) {
      case 0:
      case "APPROVED":
        return ApprovalState.APPROVED;
      case 1:
      case "PENDING":
        return ApprovalState.PENDING;
      case 2:
      case "DENIED":
        return ApprovalState.DENIED;
      case -1:
      case "UNRECOGNIZED":
      default:
        return ApprovalState.UNRECOGNIZED;
    }
  },
  toJSON(object: ApprovalState): string {
    
switch (object) {
      case ApprovalState.APPROVED:
        return 0;
      case ApprovalState.PENDING:
        return 1;
      case ApprovalState.DENIED:
        return 2;
      default:
        return -1;
    }
  },
}
```

My backend is written in Go and has troubling working with data from my ts client because Go uses integers when generating enums.

**Comment by stephenh** [comment]
Hm, I was thinking that protobuf dictated the enums-in-JSON should be strings, but you're right that it says either strings or tags are supported (and we already support parsing either, which I forgot about).

So yeah, I think some sort of something like `jsonEnumValues=strings|integers` flag seems fine.

Could you submit a PR?

Thanks!

**Comment by timostamm** [comment]
Just chiming in with some points (after implementing canonical JSON in another project):

- Enum values as integers in JSON should probably a _runtime_ option, not a plugin flag. 
- The [recommended options](https://developers.google.com/protocol-buffers/docs/proto3#json_options) are: 
  - **Emit enum values as integers instead of strings**. <- what this issue is about
  - Emit fields with default values
  - Use proto field name instead of lowerCamelCase name
  - Ignore unknown fields (when reading JSON)
- **reading JSON should accept integer / string for enum** <- what this issue is about
- reading JSON must also accept proto field name 
- field option json_name must be honored in JSON read / write

**Comment by stephenh** [comment]
> reading JSON should accept integer / string for enum

We do this already, on reading:

```typescript
export function stateEnumFromJSON(object: any): StateEnum {
  switch (object) {
    case 0:
    case "UNKNOWN":
      return StateEnum.UNKNOWN;
    case 2:
    case "ON":
      return StateEnum.ON;
    case 3:
    case "OFF":
      return StateEnum.OFF;
    case -1:
    case "UNRECOGNIZED":
    default:
      return StateEnum.UNRECOGNIZED;
  }
}
```

So I think this issue is about only "Emit enum values as integers instead of strings".

This should actually be a really easy change, assuming it's a build-time config option...

Supporting protobuf field names and `json_name` is interesting, but would be a separate enhancement and little more involved.

**Comment by hijamoya** [comment]
Any progress?

**Comment by stephenh** [comment]
@hijamoya nope, would be great if you're interested in making a PR; the code is around here:

https://github.com/stephenh/ts-proto/blob/main/src%2Fenums.ts#L115

**Comment by kev0960** [comment]
I created a pull request for this.

https://github.com/stephenh/ts-proto/pull/625

**Comment by stephenh** [comment]
Fixed in #625, thanks @kev0960 !
